src/ints/mouse_vmware.cpp

```diff
@@ -0,0 +1,247 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "mouse.h"
+
+#include <algorithm>
+
+#include "bitops.h"
+#include "checks.h"
+#include "regs.h"
+#include "inout.h"
+#include "video.h"
+
+using namespace bit::literals;
+
+CHECK_NARROWING();
+
+// VMware mouse interface passes both absolute mouse position and button
+// state to the guest side driver, but still relies on PS/2 interface,
+// which has to be used to listen for events
+
+// Reference:
+// - https://wiki.osdev.org/VMware_tools
+// - https://wiki.osdev.org/VirtualBox_Guest_Additions (planned support)
+// Drivers:
+// - https://github.com/NattyNarwhal/vmwmouse
+// - https://git.javispedro.com/cgit/vbmouse.git (planned support)
+
+enum VMwareCmd : uint16_t {
+	GetVersion        = 10,
+	AbsPointerData    = 39,
+	AbsPointerStatus  = 40,
+	AbsPointerCommand = 41,
+};
+
+enum VMwareAbsPointer : uint32_t {
+	Enable   = 0x45414552,
+	Relative = 0xF5,
+	Absolute = 0x53424152,
+};
+
+static constexpr io_port_t VMWARE_PORT = 0x5658u; // communication port
+// static constexpr io_port_t VMWARE_PORTHB = 0x5659u;  // communication port,
+// high bandwidth
+static constexpr uint32_t VMWARE_MAGIC = 0x564D5868u; // magic number for all
+                                                      // VMware calls
+
+static bool updated = false; // true = mouse state update waits to be piced up
+static uint8_t buttons_vmware = 0; // state of mouse buttons, in VMware format
+static uint16_t scaled_x = 0x7fff; // absolute mouse position, scaled from 0 to
+                                   // 0xffff
+static uint16_t scaled_y = 0x7fff; // 0x7fff is a center position
+static int8_t wheel      = 0;      // wheel movement counter
+
+static int16_t offset_x = 0; // offses between host and guest mouse coordinates
+                             // (in host pixels)
+static int16_t offset_y = 0; // (in host pixels)
+
+bool mouse_vmware = false; // if true, VMware compatible driver has taken over
+                           // the mouse
+
+// ***************************************************************************
+// VMware interface implementation
+// ***************************************************************************
+
+static void CmdGetVersion()
+{
+	reg_eax = 0; // TODO: should we respond with something resembling
+	             // VMware? For now 0 seems OK
+	reg_ebx = VMWARE_MAGIC;
+}
+
+static void CmdAbsPointerData()
+{
+	reg_eax = buttons_vmware;
+	reg_ebx = scaled_x;
+	reg_ecx = scaled_y;
+	reg_edx = static_cast<uint32_t>((wheel >= 0) ? wheel : 0x100 + wheel);
+
+	wheel = 0;
+}
+
+static void CmdAbsPointerStatus()
+{
+	reg_eax = updated ? 4 : 0;
```

Comment:
Constant with descriptive name for 4 and 0?

Reply:
Done.